### PR TITLE
timestamp add And Sub support for Snowflake

### PR DESCRIPTION
**Pull request by rohitgaurDatametica**

### CODE REVIEW

**Review by pratik163 — APPROVED** [review]

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 1]
```diff
@@ -702,6 +703,31 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
       call.operand(1).unparse(writer, leftPrec, rightPrec);
       writer.endFunCall(frame);
       break;
+    case "TIMESTAMPINTMUL":
+      if (call.getOperandList().get(0).getClass().getName()
```
Comment (deepakDatametica) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:707]: please create separate method for this logic

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 2]
```diff
@@ -702,6 +703,31 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
...
+              .equals("org.apache.calcite.sql.SqlBasicCall")) {
```
Comment (deepakDatametica) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:708]: check call.operand(0) instance of SqlBasicCall. inspite of checking with className

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 3]
```diff
@@ -702,6 +703,31 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
...
+        handleSqlBasicCallForTimestampMulti(writer, call);
+      } else {
+        SqlIntervalLiteral intervalLiteralValue = (SqlIntervalLiteral) call.getOperandList().get(0);
+        SqlIntervalLiteral.IntervalValue literalValue =
+                (SqlIntervalLiteral.IntervalValue) intervalLiteralValue.getValue();
+        String secondOperand = "";
+        if (call.getOperandList().get(1).getClass().getName()
```
Comment (deepakDatametica) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:715]: same here. check with instanceof

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 4]
```diff
@@ -431,6 +435,26 @@ private SqlNode getIdentifier(SqlBasicCall intervalOperand) {
...
+  private void unparseTimestamp(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    if (call.getOperator().toString().equals("TIMESTAMP_ADD")
+            || call.getOperator().toString().equals("TIMESTAMP_SUB")) {
+      final SqlWriter.Frame formatTime = writer.startFunCall("TIMESTAMPADD");
+      String formatCall = ((SqlIntervalLiteral) call.operand(1))
+              .getTypeName().toString().split("_")[1];
+      writer.print(formatCall + ", ");
+      String intCall = ((SqlIntervalLiteral) call.operand(1)).getValue().toString();
+      if (call.getOperator().toString().equals("TIMESTAMP_SUB")) {
+        writer.print("-" + intCall + ", ");
```
Comment (deepakDatametica) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:447]: `if (true) {
writer.print("-"); 
}
writer.print(intCall + ", ")
`
and remove else part

File: core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java [hunk 5]
```diff
@@ -636,6 +640,23 @@ private SqlCharStringLiteral creteDateTimeFormatSqlCharLiteral(String format) {
       String standardDateFormat, Map<SqlDateTimeFormat, String> dateTimeFormatMap) {
     return super.getDateTimeFormatString(standardDateFormat, dateTimeFormatMap);
   }
+
+  private void unparsePlusMinus(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    switch (call.getOperator().getName()) {
+    case "TIMESTAMP_ADD":
+      call.operand(0).unparse(writer, leftPrec, rightPrec);
+      writer.print("+ ");
```
Comment (deepakDatametica) [inline core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java:648]: create a separate method for this and pass `sign` "+" or "-" as argument to that method.

File: core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java [hunk 6]
```diff
@@ -678,5 +682,26 @@ private SqlCharStringLiteral creteDateTimeFormatSqlCharLiteral(String format) {
     return super.getDateTimeFormatString(standardDateFormat, dateTimeFormatMap);
   }
 
+  private void unparsePlusMinus(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    switch (call.getOperator().getName()) {
+    case "TIMESTAMP_ADD":
+    case "TIMESTAMP_SUB":
+      call.operand(0).unparse(writer, leftPrec, rightPrec);
```
Comment (pratik163) [inline core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java:689]: do we need to add operator before unparsing?

File: core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java [hunk 7]
```diff
@@ -636,6 +640,27 @@ private SqlCharStringLiteral creteDateTimeFormatSqlCharLiteral(String format) {
...
+      checkSign(writer, call);
+      call.operand(1).unparse(writer, leftPrec, rightPrec);
+      break;
+    default:
+      super.unparseCall(writer, call, leftPrec, rightPrec);
+    }
+  }
+
+  private void checkSign(SqlWriter writer, SqlCall call) {
+    if (call.getKind() == SqlKind.PLUS) {
```
Comment (pratik163) [inline core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java:658]: switch known value to left side of comparison

File: core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java (hunk 6 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java:689]: no, we don't need to add the operator in hive and spark

File: core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java (hunk 7 above)
Comment (rohitgaurDatametica) [reply core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java:658]: done

**Review by pratik163 — APPROVED** [review]

**Review by deepakDatametica — APPROVED** [review]

**Pull request merged**